docs/langref/relay_pattern.rst

```diff
@@ -0,0 +1,143 @@
+
+The next example is a dense operation with any operator that is marked element-wise::
+
+    def test_no_match_attr():
+        op = is_op('nn.dense').has_attr("TOpPattern", K_ELEMWISE)
+        op_pat = op(wildcard(), wildcard())
+        x = relay.var('x')
+        y = relay.var('y')
+        assert not op_pat.match(relay.op.nn.dense(x, y))
+
+The next example is matching a diamond with two inputs at the top of the diamond::
+
+    def test_match_diamond():
+        # Pattern
+        is_conv2d = is_op('nn.conv2d')(is_input(), is_input())
+        path1 = is_op('nn.relu')(is_conv2d)
+        path2 = is_op('nn.leaky_relu')(is_conv2d)
+        diamond = is_op('add')(path1, path2)
+
+        # Expr
+        inp = relay.var('input')
+        weight = relay.var('weight')
+        conv2d = relay.op.nn.conv2d(inp, weight)
+        relu = relay.op.nn.relu(conv2d)
+        leaky_relu = relay.op.nn.leaky_relu(conv2d, alpha=0)
+        out = relu + leaky_relu
+
+        # Check
+        assert diamond.match(out)
+
+The final example we would like to match which is not yet implemented in the prototype is matching diamonds with a post-dominator relationship. Our plan is to embed dominator analysis as type of matching in the pattern language in order to allow for pattern matching with unknown topology.  This is important because we want to able to use the language to describe fuse patterns, like elementwise operations followed by a conv2d::
+
+    def test_match_dom_diamond():
+        # Pattern
+        is_conv2d = is_op('nn.conv2d')(is_input(), is_input())
+        reduction = is_op('add')(wildcard(), wildcard())
+    	diamond = dominates(is_conv2d, is_elemwise, reduction)
+
+        # Expr
+        inp = relay.var('input')
+        weight = relay.var('weight')
+        conv2d = relay.op.nn.conv2d(inp, weight)
+        relu = relay.op.nn.relu(conv2d)
+        leaky_relu = relay.op.nn.leaky_relu(conv2d, alpha=0)
+        out = relu + leaky_relu
+
+        # Check
+        assert diamond.match(out)
+
+Design
+======
+
+The pattern language proposed is designed to be a mirror of Relay's IR with additional support for common scenarios. The goal of the pattern language is to provide a regular-expression like capability for matching data-flow graphs and doing rewriting.
+
+The high level design is to introduce a language of patterns for now we propose the language as::
+
+    Pattern ::= expr
+            | *
+            | pattern(pattern1, ... patternN)
+            | has_type(pattern, type)
+            | has_attr(pattern, attr, attr_value)
+            | is_input(name)
+            | pattern1 `|` pattern2
+            | dominates(parent_pattern, path_pattern, child_pattern)
+
+The above language then provides a matching interface with both can select sub-graphs as well as verify that the graph does match the pattern.
+
+Expression Pattern
+******************
+
+Match a literal expression.
+
+Wildcard
+******************
```

Comment:
restructuredText usually requires the section annotation line to have the same length as the above title.

```
Wildcard
**********
```